No known way to pull from an insecure registry within a jib build

when trying to pull an image from an insecure registry with JIB strategy an exception is raised: 

Failed to verify the server at https://harbor.local/v2/jkube/jkube-tomcat9-binary-s2i/manifests/0.0.5 because only secure connections are allowed.

There is no (documented) way to specify e. g. `allowInsecureRegistries` (the way it is possible with jib-maven-plugin).

### Info

* Eclipse JKube version : 1.0.0-rc-1
* Maven version (`mvn -v`) : 3.6.3
* Kubernetes / Red Hat OpenShift setup and version : Rancher 2.4.2, K8s 1.17.4


Create a small project with the following pom.xml snippet an do a maven k8s:build. if the registry uses a self signed certificate or a custom certificate the exception is rased.
```
<plugin>
	<groupId>org.eclipse.jkube</groupId>
	<artifactId>kubernetes-maven-plugin</artifactId>
	<version>1.0.0-rc-1</version>
	<configuration>
		<buildStrategy>jib</buildStrategy>
		<images>
			<image>
				<name>library/${project.artifactId}:${project.version}</name>
				<registry>harbor.local</registry>
				<build>
					<from>harbor.local/jkube/jkube-tomcat9-binary-s2i:0.0.5</from>
				</build>
				...
			</image>
		</images>
	</configuration>
</plugin>
```

Error log when reproducing issue:
```shell
JIB> Base image 'localhost:5000/jkube/jkube' does not use a specific image digest - build may not be reproducible            
JIB> Containerizing application with the following files:                                                                    
JIB>    :                                                                                                                      
JIB>            /home/user/00-MN/projects/forks/jkube/quickstarts/maven/spring-boot-with-jib/target/docker/rohankanojia/spring-boot-sample/build/deployments
JIB>            /home/user/00-MN/projects/forks/jkube/quickstarts/maven/spring-boot-with-jib/target/docker/rohankanojia/spring-boot-sample/build/deployments/static
JIB>            /home/user/00-MN/projects/forks/jkube/quickstarts/maven/spring-boot-with-jib/target/docker/rohankanojia/spring-boot-sample/build/deployments/static/testFile.txt
JIB>            /home/user/00-MN/projects/forks/jkube/quickstarts/maven/spring-boot-with-jib/target/docker/rohankanojia/spring-boot-sample/build/deployments/eclipse-jkube-sample-spring-boot-jib-1.0.0-rc-1.jar
JIB>    :                                                                                                                      
JIB>            /home/user/00-MN/projects/forks/jkube/quickstarts/maven/spring-boot-with-jib/target/docker/rohankanojia/spring-boot-sample/build/Dockerfile
JIB> Getting manifest for base image localhost:5000/jkube/jkube...                                                           
JIB> Building  layer...                                                                                                      
JIB> Building  layer...                                                                                                      
JIB> [=========                     ] 30.0% complete > building  layer
[ERROR] k8s: Unable to build the image tarball: : com.google.cloud.tools.jib.api.InsecureRegistryException: Failed to verify the server at https://localhost:5000/v2/jkube/jkube/manifests/latest because only secure connections are alJIB> [============                  ] 40.0% complete > building image to tar file
[ERROR] k8s: Failed to execute the build [Error when building JIB image]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.820 s
[INFO] Finished at: 2020-08-05T12:32:43+02:00
[INFO] ------------------------------------------------------------------------
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,12 +21,13 @@ Usage:
 ./scripts/extract-changelog-for-version.sh 1.3.37 5
 ```
 ### 1.0.0-SNAPSHOT
-* Doc: Fixed documentation for jkube.openshift.imageChangeTriggers
+* Fix #344: Fix documentation for `jkube.openshift.imageChangeTriggers`
 * Fix #290: Bump Fabric8 Kubernetes Client to v4.10.3
 * Fix #273: Added new parameter to allow for custom _app_ name
 * Fix #329: Vert.x generator and enricher applicable when `io.vertx` dependencies present
 * Fix #334: Build strategy picked up from XML configuration and properties
-* Fix: Maven Quickstart for Spring-Boot with complete Camel integration
+* Fix #342: Maven Quickstart for Spring-Boot with complete Camel integration
+* Fix #336: Fix pull from insecure registries and pull registry authentication for JIB
 
 ### 1.0.0-rc-1 (2020-07-23)
 * Fix #252: Replace Quarkus Native Base image with ubi-minimal (same as in `Dockerfile.native`)

--- a/jkube-kit/build/service/jib/pom.xml
+++ b/jkube-kit/build/service/jib/pom.xml
@@ -52,6 +52,10 @@
       <artifactId>junit</artifactId>
     </dependency>
     <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+    </dependency>
+    <dependency>
       <groupId>org.hamcrest</groupId>
       <artifactId>hamcrest-all</artifactId>
     </dependency>

--- a/jkube-kit/build/service/jib/src/main/java/org/eclipse/jkube/kit/service/jib/JibServiceUtil.java
+++ b/jkube-kit/build/service/jib/src/main/java/org/eclipse/jkube/kit/service/jib/JibServiceUtil.java
@@ -98,9 +98,10 @@ public class JibServiceUtil {
         try {
             jibContainerBuilder.setCreationTime(Instant.now());
             jibContainerBuilder.containerize(Containerizer.to(image)
-                    .setExecutorService(jibBuildExecutor)
-                    .addEventHandler(LogEvent.class, log(logger))
-                    .addEventHandler(ProgressEvent.class, new ProgressEventHandler(logUpdate())));
+                .setAllowInsecureRegistries(true)
+                .setExecutorService(jibBuildExecutor)
+                .addEventHandler(LogEvent.class, log(logger))
+                .addEventHandler(ProgressEvent.class, new ProgressEventHandler(logUpdate())));
             logUpdateFinished();
         } catch (CacheDirectoryCreationException | IOException | ExecutionException | RegistryException ex) {
             logger.error("Unable to build the image tarball: ", ex);
@@ -114,10 +115,9 @@ public class JibServiceUtil {
         }
     }
 
-    public static JibContainerBuilder containerFromImageConfiguration(ImageConfiguration imageConfiguration)
-            throws InvalidImageReferenceException {
-
-        final JibContainerBuilder containerBuilder = Jib.from(getBaseImage(imageConfiguration))
+    public static JibContainerBuilder containerFromImageConfiguration(
+        ImageConfiguration imageConfiguration, Credential pullRegistryCredential) throws InvalidImageReferenceException {
+        final JibContainerBuilder containerBuilder = Jib.from(getRegistryImage(getBaseImage(imageConfiguration), pullRegistryCredential))
                 .setFormat(ImageFormat.OCI);
         return populateContainerBuilderFromImageConfiguration(containerBuilder, imageConfiguration);
     }
@@ -158,7 +158,7 @@ public class JibServiceUtil {
         }
     }
 
-    protected static void pushImage(TarImage baseImage, String targetImageName, Credential credential, KitLogger logger)
+    private static void pushImage(TarImage baseImage, String targetImageName, Credential credential, KitLogger logger)
             throws InterruptedException {
 
         final ExecutorService jibBuildExecutor = Executors.newCachedThreadPool();
@@ -176,7 +176,7 @@ public class JibServiceUtil {
         }
     }
 
-    protected static JibContainerBuilder populateContainerBuilderFromImageConfiguration(JibContainerBuilder containerBuilder, ImageConfiguration imageConfiguration) {
+    private static JibContainerBuilder populateContainerBuilderFromImageConfiguration(JibContainerBuilder containerBuilder, ImageConfiguration imageConfiguration) {
         final Optional<BuildConfiguration> bic =
                 Optional.ofNullable(Objects.requireNonNull(imageConfiguration).getBuildConfiguration());
         bic.map(BuildConfiguration::getEntryPoint)
@@ -211,7 +211,7 @@ public class JibServiceUtil {
         return containerBuilder;
     }
 
-    protected static Set<String> getAllImageTags(List<String> tags, String imageName) {
+    static Set<String> getAllImageTags(List<String> tags, String imageName) {
         ImageName tempImage = new ImageName(imageName);
         Set<String> tagSet = tags.stream().filter(Objects::nonNull).collect(Collectors.toSet());
         if (!tempImage.getTag().isEmpty()) {
@@ -220,7 +220,7 @@ public class JibServiceUtil {
         return tagSet;
     }
 
-    protected static void submitPushToJib(TarImage baseImage, RegistryImage targetImage, ExecutorService jibBuildExecutor, KitLogger logger) throws InterruptedException, ExecutionException, RegistryException, CacheDirectoryCreationException, IOException {
+    private static void submitPushToJib(TarImage baseImage, RegistryImage targetImage, ExecutorService jibBuildExecutor, KitLogger logger) throws InterruptedException, ExecutionException, RegistryException, CacheDirectoryCreationException, IOException {
         Jib.from(baseImage).containerize(Containerizer.to(targetImage)
                 .setExecutorService(jibBuildExecutor)
                 .addEventHandler(LogEvent.class, log(logger))
@@ -228,7 +228,7 @@ public class JibServiceUtil {
         logUpdateFinished();
     }
 
-    protected static RegistryImage getRegistryImage(String targetImage, Credential credential) throws InvalidImageReferenceException {
+    private static RegistryImage getRegistryImage(String targetImage, Credential credential) throws InvalidImageReferenceException {
         RegistryImage registryImage = RegistryImage.named(targetImage);
         if (credential != null && !credential.getUsername().isEmpty() && !credential.getPassword().isEmpty()) {
             registryImage.addCredential(credential.getUsername(), credential.getPassword());
@@ -339,7 +339,7 @@ public class JibServiceUtil {
      * @param unfinishedLeafTasks the unfinished leaf tasks
      * @return the progress display as a list of lines
      */
-    public static List<String> generateProgressDisplay(double progress, List<String> unfinishedLeafTasks) {
+    private static List<String> generateProgressDisplay(double progress, List<String> unfinishedLeafTasks) {
         List<String> lines = new ArrayList<>();
 
         lines.add(HEADER);

--- a/jkube-kit/build/service/jib/src/test/java/org/eclipse/jkube/kit/service/jib/JibServiceUtilTest.java
+++ b/jkube-kit/build/service/jib/src/test/java/org/eclipse/jkube/kit/service/jib/JibServiceUtilTest.java
@@ -13,20 +13,6 @@
  */
 package org.eclipse.jkube.kit.service.jib;
 
-import com.google.cloud.tools.jib.api.AbsoluteUnixPath;
-import com.google.cloud.tools.jib.api.JibContainerBuilder;
-import com.google.cloud.tools.jib.api.LayerConfiguration;
-import com.google.cloud.tools.jib.api.Port;
-import mockit.Mocked;
-import mockit.Verifications;
-import org.eclipse.jkube.kit.common.Assembly;
-import org.eclipse.jkube.kit.common.AssemblyConfiguration;
-import org.eclipse.jkube.kit.common.AssemblyFile;
-import org.eclipse.jkube.kit.config.image.ImageConfiguration;
-import org.eclipse.jkube.kit.config.image.build.Arguments;
-import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
-import org.junit.Test;
-
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
@@ -36,7 +22,24 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
+import com.google.cloud.tools.jib.api.ImageFormat;
+import org.eclipse.jkube.kit.common.Assembly;
+import org.eclipse.jkube.kit.common.AssemblyConfiguration;
+import org.eclipse.jkube.kit.common.AssemblyFile;
+import org.eclipse.jkube.kit.config.image.ImageConfiguration;
+import org.eclipse.jkube.kit.config.image.build.Arguments;
+import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
+
+import com.google.cloud.tools.jib.api.AbsoluteUnixPath;
+import com.google.cloud.tools.jib.api.JibContainerBuilder;
+import com.google.cloud.tools.jib.api.LayerConfiguration;
+import com.google.cloud.tools.jib.api.Port;
+import mockit.Mocked;
+import mockit.Verifications;
+import org.junit.Test;
+
 import static org.eclipse.jkube.kit.service.jib.JibServiceUtil.BUSYBOX;
+import static org.eclipse.jkube.kit.service.jib.JibServiceUtil.containerFromImageConfiguration;
 import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
@@ -58,29 +61,24 @@ public class JibServiceUtilTest {
     }
 
     @Test
-    public void testContainerFromImageConfiguration(@Mocked JibContainerBuilder containerBuilder) {
+    public void testContainerFromImageConfiguration(@Mocked JibContainerBuilder containerBuilder)  throws Exception{
         // Given
         ImageConfiguration imageConfiguration = getSampleImageConfiguration();
-
         // When
-        JibContainerBuilder jibContainerBuilder = JibServiceUtil.populateContainerBuilderFromImageConfiguration(containerBuilder, imageConfiguration);
-
+        JibContainerBuilder jibContainerBuilder = containerFromImageConfiguration(imageConfiguration, null);
         // Then
-        assertNotNull(jibContainerBuilder);
+        // @formatter:off
         new Verifications() {{
-            jibContainerBuilder.addLabel("foo", "bar");
-            times = 1;
-            jibContainerBuilder.setEntrypoint(Arrays.asList("java", "-jar", "foo.jar"));
-            times = 1;
-            jibContainerBuilder.setExposedPorts(new HashSet<>(Collections.singletonList(Port.tcp(8080))));
-            times = 1;
-            jibContainerBuilder.setUser("root");
-            times = 1;
-            jibContainerBuilder.setWorkingDirectory(AbsoluteUnixPath.get("/home/foo"));
-            times = 1;
+            jibContainerBuilder.addLabel("foo", "bar"); times = 1;
+            jibContainerBuilder.setEntrypoint(Arrays.asList("java", "-jar", "foo.jar")); times = 1;
+            jibContainerBuilder.setExposedPorts(new HashSet<>(Collections.singletonList(Port.tcp(8080)))); times = 1;
+            jibContainerBuilder.setUser("root"); times = 1;
+            jibContainerBuilder.setWorkingDirectory(AbsoluteUnixPath.get("/home/foo")); times = 1;
             jibContainerBuilder.setVolumes(new HashSet<>(Collections.singletonList(AbsoluteUnixPath.get("/mnt/volume1"))));
             times = 1;
+            jibContainerBuilder.setFormat(ImageFormat.OCI); times = 1;
         }};
+        // @formatter:on
     }
 
     @Test

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildService.java
@@ -45,14 +45,17 @@ import java.util.Objects;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
+import static org.eclipse.jkube.kit.service.jib.JibServiceUtil.containerFromImageConfiguration;
+import static org.eclipse.jkube.kit.service.jib.JibServiceUtil.getBaseImage;
+
 public class JibBuildService implements BuildService {
     private static final String DOCKER_LOGIN_DEFAULT_REGISTRY = "https://index.docker.io/v1/";
     private static final List<String> DEFAULT_DOCKER_REGISTRIES = Arrays.asList(
             "docker.io", "index.docker.io", "registry.hub.docker.com"
     );
     private static final String PUSH_REGISTRY = "jkube.docker.push.registry";
-    private JKubeServiceHub jKubeServiceHub;
-    private KitLogger log;
+    private final JKubeServiceHub jKubeServiceHub;
+    private final KitLogger log;
 
     public JibBuildService(JKubeServiceHub jKubeServiceHub, KitLogger logger) {
         Objects.requireNonNull(jKubeServiceHub.getBuildServiceConfig(), "BuildServiceConfig is required");
@@ -68,9 +71,11 @@ public class JibBuildService implements BuildService {
             if (imageConfig.getBuildConfiguration().isDockerFileMode()) {
                 throw new JKubeServiceException("Dockerfile mode is not supported with JIB build strategy");
             }
-            appendRegistry(imageConfig, configuration.getProperties().getProperty(PUSH_REGISTRY));
+            prependRegistry(imageConfig, configuration.getProperties().getProperty(PUSH_REGISTRY));
             BuildDirs buildDirs = new BuildDirs(imageConfig.getName(), configuration);
-            final JibContainerBuilder containerBuilder = JibServiceUtil.containerFromImageConfiguration(imageConfig);
+            final Credential pullRegistryCredential = getRegistryCredentials(
+                configuration.getRegistryConfig(), false, imageConfig, log);
+            final JibContainerBuilder containerBuilder = containerFromImageConfiguration(imageConfig, pullRegistryCredential);
 
             File dockerTarArchive = getAssemblyTarArchive(imageConfig, configuration, log);
             // TODO: Improve Assembly Manager so that the effective assemblyFileEntries computed can be properly shared
@@ -99,9 +104,14 @@ public class JibBuildService implements BuildService {
     public void push(Collection<ImageConfiguration> imageConfigs, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
         try {
             for (ImageConfiguration imageConfiguration : imageConfigs) {
-                appendRegistry(imageConfiguration, registryConfig.getRegistry());
+                prependRegistry(imageConfiguration, registryConfig.getRegistry());
                 log.info("This push refers to: %s", imageConfiguration.getName());
-                JibServiceUtil.jibPush(imageConfiguration, getRegistryCredentials(registryConfig, imageConfiguration, log), getBuildTarArchive(imageConfiguration, jKubeServiceHub), log);
+                JibServiceUtil.jibPush(
+                    imageConfiguration,
+                    getRegistryCredentials(registryConfig, true, imageConfiguration, log),
+                    getBuildTarArchive(imageConfiguration, jKubeServiceHub),
+                    log
+                );
             }
         } catch (Exception ex) {
             throw new JKubeServiceException("Error when push JIB image", ex);
@@ -113,7 +123,7 @@ public class JibBuildService implements BuildService {
         // No post processing required
     }
 
-    protected static ImageConfiguration appendRegistry(ImageConfiguration imageConfiguration, String registry) {
+    static ImageConfiguration prependRegistry(ImageConfiguration imageConfiguration, String registry) {
         ImageName imageName = new ImageName(imageConfiguration.getName());
         if (!imageName.hasRegistry() && registry != null) {
             imageConfiguration.setName(registry + "/" + imageConfiguration.getName());
@@ -122,21 +132,36 @@ public class JibBuildService implements BuildService {
         return imageConfiguration;
     }
 
-    protected static File getAssemblyTarArchive(ImageConfiguration imageConfig, JKubeConfiguration configuration, KitLogger log) throws IOException {
+    static File getAssemblyTarArchive(ImageConfiguration imageConfig, JKubeConfiguration configuration, KitLogger log) throws IOException {
         log.info("Preparing assembly files");
         final String targetImage = imageConfig.getName();
         return AssemblyManager.getInstance()
                 .createDockerTarArchive(targetImage, configuration, imageConfig.getBuildConfiguration(), log, null);
     }
 
-    protected static Credential getRegistryCredentials(RegistryConfig registryConfig, ImageConfiguration imageConfiguration, KitLogger log) throws IOException {
-        String registry = EnvUtil.firstRegistryOf((new ImageName(imageConfiguration.getName())).getRegistry(), imageConfiguration.getRegistry(), registryConfig.getRegistry());
+    static Credential getRegistryCredentials(
+        RegistryConfig registryConfig, boolean isPush, ImageConfiguration imageConfiguration, KitLogger log)
+        throws IOException {
+
+        String registry;
+        if (isPush) {
+            registry = EnvUtil.firstRegistryOf(
+                new ImageName(imageConfiguration.getName()).getRegistry(),
+                imageConfiguration.getRegistry(),
+                registryConfig.getRegistry()
+            );
+        } else {
+            registry = EnvUtil.firstRegistryOf(
+                new ImageName(getBaseImage(imageConfiguration)).getRegistry(),
+                registryConfig.getRegistry()
+            );
+        }
         if (registry == null || DEFAULT_DOCKER_REGISTRIES.contains(registry)) {
             registry = DOCKER_LOGIN_DEFAULT_REGISTRY; // Let's assume docker is default registry.
         }
 
         AuthConfigFactory authConfigFactory = new AuthConfigFactory(log);
-        AuthConfig standardAuthConfig = authConfigFactory.createAuthConfig(true, registryConfig.isSkipExtendedAuth(), registryConfig.getAuthConfig(), registryConfig.getSettings(), null, registry, registryConfig.getPasswordDecryptionMethod());
+        AuthConfig standardAuthConfig = authConfigFactory.createAuthConfig(isPush, registryConfig.isSkipExtendedAuth(), registryConfig.getAuthConfig(), registryConfig.getSettings(), null, registry, registryConfig.getPasswordDecryptionMethod());
         Credential credentials = null;
         if (standardAuthConfig != null) {
             credentials = Credential.from(standardAuthConfig.getUsername(), standardAuthConfig.getPassword());
@@ -144,7 +169,7 @@ public class JibBuildService implements BuildService {
         return credentials;
     }
 
-    protected static File getBuildTarArchive(ImageConfiguration imageConfiguration, JKubeServiceHub jKubeServiceHub) {
+    static File getBuildTarArchive(ImageConfiguration imageConfiguration, JKubeServiceHub jKubeServiceHub) {
         BuildDirs buildDirs = new BuildDirs(imageConfiguration.getName(), jKubeServiceHub.getConfiguration());
         return new File(buildDirs.getTemporaryRootDirectory(), JKubeBuildTarArchiver.ARCHIVE_FILE_NAME + ArchiveCompression.none.getFileSuffix());
     }

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildServiceTest.java
@@ -18,6 +18,7 @@ import mockit.Expectations;
 import mockit.Mock;
 import mockit.MockUp;
 import mockit.Mocked;
+import mockit.Verifications;
 import org.eclipse.jkube.kit.build.api.auth.AuthConfig;
 import org.eclipse.jkube.kit.build.service.docker.auth.AuthConfigFactory;
 import org.eclipse.jkube.kit.common.JavaProject;
@@ -28,6 +29,7 @@ import org.eclipse.jkube.kit.config.image.RegistryConfig;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 import org.eclipse.jkube.kit.config.image.build.JKubeConfiguration;
 import org.eclipse.jkube.kit.config.service.JKubeServiceHub;
+import org.eclipse.jkube.kit.service.jib.JibServiceUtil;
 import org.junit.Test;
 
 import java.io.File;
@@ -44,37 +46,46 @@ import static org.junit.Assert.assertNotNull;
 
 public class JibBuildServiceTest {
     @Mocked
-    KitLogger logger;
+    private KitLogger logger;
 
     @Mocked
-    JKubeServiceHub serviceHub;
+    private JKubeServiceHub serviceHub;
 
     @Test
     @java.lang.SuppressWarnings("squid:S00112")
-    public void testGetRegistryCredentials() throws IOException {
+    public void testGetRegistryCredentialsForPush() throws IOException {
         // Given
         ImageConfiguration imageConfiguration = getImageConfiguration();
-        new MockUp<AuthConfigFactory>() {
-            @Mock
-            AuthConfig createStandardAuthConfig(boolean isPush, Map authConfigMap, List<RegistryServerConfiguration> settings, String user, String registry, UnaryOperator<String> passwordDecryptionMethod, KitLogger log)
-                    throws IOException {
-                return AuthConfig.builder()
-                        .username("testuser")
-                        .password("testpass")
-                        .build();
-            }
-        };
+        mockAuthConfig();
         RegistryConfig registryConfig = RegistryConfig.builder()
                 .authConfig(Collections.emptyMap())
                 .settings(Collections.emptyList())
                 .build();
-
         // When
-        Credential credential = JibBuildService.getRegistryCredentials(registryConfig, imageConfiguration, logger);
-
+        Credential credential = JibBuildService.getRegistryCredentials(
+            registryConfig, true, imageConfiguration, logger);
         // Then
         assertNotNull(credential);
-        assertEquals("testuser", credential.getUsername());
+        assertEquals("testuserpush", credential.getUsername());
+        assertEquals("testpass", credential.getPassword());
+    }
+
+    @Test
+    @java.lang.SuppressWarnings("squid:S00112")
+    public void testGetRegistryCredentialsForPull() throws IOException {
+        // Given
+        ImageConfiguration imageConfiguration = getImageConfiguration();
+        mockAuthConfig();
+        RegistryConfig registryConfig = RegistryConfig.builder()
+            .authConfig(Collections.emptyMap())
+            .settings(Collections.emptyList())
+            .build();
+        // When
+        Credential credential = JibBuildService.getRegistryCredentials(
+            registryConfig, false, imageConfiguration, logger);
+        // Then
+        assertNotNull(credential);
+        assertEquals("testuserpull", credential.getUsername());
         assertEquals("testpass", credential.getPassword());
     }
 
@@ -110,22 +121,54 @@ public class JibBuildServiceTest {
     }
 
     @Test
-    public void testAppendRegistry() {
+    public void testPrependRegistry() {
         // Given
         ImageConfiguration imageConfiguration = getImageConfiguration();
-
         // When
-        JibBuildService.appendRegistry(imageConfiguration, "quay.io");
-
+        JibBuildService.prependRegistry(imageConfiguration, "quay.io");
         // Then
         assertNotNull(imageConfiguration);
         assertEquals("quay.io/test/testimage:0.0.1", imageConfiguration.getName());
     }
 
+    @Test
+    public void testPushWithNoConfigurations(@Mocked JibServiceUtil jibServiceUtil) throws Exception {
+        // When
+        new JibBuildService(serviceHub, logger).push(Collections.emptyList(), 1, null, false);
+        // Then
+        // @formatter:off
+        new Verifications() {{
+            JibServiceUtil.jibPush((ImageConfiguration)any, (Credential)any, (File)any, logger); times = 0;
+        }};
+        // @formatter:on
+    }
+
+    @Test
+    public void testPushWithConfiguration(@Mocked JibServiceUtil jibServiceUtil) throws Exception {
+        // Given
+        mockAuthConfig();
+        final ImageConfiguration imageConfiguration = getImageConfiguration();
+        final RegistryConfig registryConfig = RegistryConfig.builder()
+            .build();
+        // When
+        new JibBuildService(serviceHub, logger).push(Collections.singletonList(imageConfiguration), 1, registryConfig, false);
+        // Then
+        // @formatter:off
+        new Verifications() {{
+            JibServiceUtil.jibPush(
+                imageConfiguration,
+                Credential.from("testuserpush", "testpass"),
+                (File)any,
+                logger);
+            times = 1;
+        }};
+        // @formatter:on
+    }
+
     private ImageConfiguration getImageConfiguration() {
         return ImageConfiguration.builder()
                 .name("test/testimage:0.0.1")
-                .build(BuildConfiguration.builder().build())
+                .build(BuildConfiguration.builder().from("busybox").build())
                 .build();
     }
 
@@ -141,5 +184,19 @@ public class JibBuildServiceTest {
             serviceHub.getConfiguration();
             result = jKubeConfiguration;
         }};
+    }
+
+
+    private static void mockAuthConfig() {
+        new MockUp<AuthConfigFactory>() {
+            @Mock
+            AuthConfig createStandardAuthConfig(boolean isPush, Map authConfigMap, List<RegistryServerConfiguration> settings, String user, String registry, UnaryOperator<String> passwordDecryptionMethod, KitLogger log)
+                throws IOException {
+                return AuthConfig.builder()
+                    .username("testuser" + (isPush ? "push" : "pull"))
+                    .password("testpass")
+                    .build();
+            }
+        };
     }
 }

